host_cleanup.sh

```diff
@@ -31,6 +31,8 @@ if [ "${CAPM3_RUN_LOCAL}" = true ]; then
   fi
 fi
 
+# Clean up the local container registry
+sudo "${CONTAINER_RUNTIME}" rm registry -f || true
```

Comment:
I think this won't be enough to solve the problem we have in the Bare Metal Lab, because the any images that have been pulled from the internet to Docker's image storage won't be cleaned, and on next run, `pull-if-not-present` will skip the pull, and the same old images will be pushed to the registry again. I expect you may need to do a `docker image prune` here as well.

The history of this is, we once had a lot of CI failures due to hitting Docker's API rate limit, and one of the things I tried to mitigate is to ensure we didn't pull any images during CI runs, but only in CI VM image build. My unwise idea was to not attempt `docker pull` if we already had the image matching the tag we need. Then when we build the image and run the 01 script there on a fresh VM, we pull the images, but in CI run, we have the images already, so no matter how many times you run the build, no `docker pull` is done. Hence `pull-if-not-present`, and this works to a point. Unfortunately, the problem is that a lot of the images are tagged `latest`: then if updates are pushed upstream, the new function doesn't bother to try to pull them, since it sees it already has a matching image named :latest. In long running environments like the BML, we end up with stale 'latest' images locally, which are then kept in the local registry.

The real fix to the problem above is to use a registry which caches the images and get rid of `pull-if-not-present`. Then we can do a pull, get the freshest image if it exists every time and not hit a rate limit. @furkatgofurov7 had success setting this up for Quay, so this may already exist and I think it's the best way to solve both the rate limit issue and to make sure we have the latest images in every environment.

The other solution to fix the Bare Metal Lab is to add image prune and registry teardown to the Bare Metal Lab specific playbook: https://github.com/metal3-io/project-infra/blob/d369c5a96c31d7c8e73b73d45dc7959d4323c346/jenkins/scripts/bare_metal_lab/deploy-lab.yaml#L59

Reply:
Oh I see. Thank you @macaptain ! I'll investigate if we can use Quay.
I would argue that we should clean the local registry always when running `make clean` anyway, and I'll add pruning to this. Or do you mean that this should only be done in the BML environment? I guess there could be reasons to not prune and remove the local registry all the time, is this the case?
What I'm trying to say is that the BML specific playbook runs `make clean` so I thought it would make most sense to add the cleaning to the script that is run when you do this.

Reply:
I think the reason for not cleaning the registry and pruning the images is just that it's quite a lot to download each time. If you had a fresh image, the `docker push` _should_ update the registry. But as described, the images go stale because we stop pulling them from the internet once we have them the first time.

I think also we don't want to issue a full `docker image prune` in metal3-dev-env - it may delete other images unrelated to metal3. Instead, using the same loop we use to `pull` the images to instead `rm` the images.

In order of preference, I think the best way forward is:

1) Use @furkatgofurov7 's registry proxy in metal3-dev-env and delete `pull-if-not-present`
2) Fix the BML playbook to properly clean the registry and images
3) Selectively delete metal3 images in metal3-dev-env when running `make clean`

Reply:
@macaptain I missed a context here, which container image we are talking about here to pull from registry?

Reply:
@furkatgofurov7 

https://github.com/metal3-io/metal3-dev-env/pull/835
https://github.com/metal3-io/project-infra/pull/252

We're using this proxy to pull from Quay, but could we use this proxy registry in the same way for images pulled from DockerHub?

Reply:
@macaptain I think no, since proxy has been configured to cache images only from metal3 quay.io, all the images we want from DockerHub needs to be in quay.

Reply:
Thanks for the links @macaptain , this is really helpful!
It looks like Nordix already have a [Docker Hub proxy](https://wiki.nordix.org/display/DEV/Docker+Hub+Proxy) that we could probably use.

Reply:
@lentzi90 @macaptain actually there is already another docker-hub-proxy cache in [registry](https://registry.nordix.org/harbor/projects/25/repositories) for DockerHub images and that should work